File: Android/BarkCloud.Android/app/src/main/java/com/barkfluff/BarkCloud/data/cloud/AlbumRepository.kt
package com.barkfluff.BarkCloud.data.cloud

import barkcloud.files.FilesApiOuterClass.AddItemsToAlbumRequest
import barkcloud.files.FilesApiOuterClass.CreateAlbumRequest
import barkcloud.files.FilesApiOuterClass.DeleteAlbumRequest
import barkcloud.files.FilesApiOuterClass.ListAlbumItemsRequest
import barkcloud.files.FilesApiOuterClass.ListAlbumsRequest
import barkcloud.files.FilesApiOuterClass.MediaKind
import barkcloud.files.FilesApiOuterClass.RemoveItemsFromAlbumRequest
import barkcloud.files.FilesApiOuterClass.UpdateAlbumRequest
import com.barkfluff.BarkCloud.grpc.GrpcManager

/** Страница содержимого альбома с курсором пагинации. */
data class AlbumItemsPage(
    val items: List<MediaAsset>,
    val nextCursorAddedAtMillis: Long?,
    val nextCursorFileId: String,
) {
    val hasMore: Boolean get() = nextCursorAddedAtMillis != null
}

/** Доступ к сервису альбомов (`AlbumApi`). */
class AlbumRepository(private val grpc: GrpcManager) {

    suspend fun listAlbums(
        limit: Int = 50,
        cursorUpdatedAtMillis: Long? = null,
        cursorAlbumId: String = "",
    ): AlbumPage {
        val req = ListAlbumsRequest.newBuilder().setLimit(limit).setCursorAlbumId(cursorAlbumId)
        if (cursorUpdatedAtMillis != null) req.setCursorUpdatedAt(cursorUpdatedAtMillis.toTimestamp())
        val resp = grpc.albumStub().listAlbums(req.build())
        return AlbumPage(
            albums = resp.albumsList.map { AlbumCard.from(it) },
            nextCursorUpdatedAtMillis = if (resp.hasNextCursorUpdatedAt()) resp.nextCursorUpdatedAt.toEpochMillis() else null,
            nextCursorAlbumId = resp.nextCursorAlbumId,
        )
    }

    suspend fun listItems(
        albumId: String,
        kindFilter: CloudMediaKind? = null,
        limit: Int = 50,
        cursorAddedAtMillis: Long? = null,
        cursorFileId: String = "",
    ): AlbumItemsPage {
        val req = ListAlbumItemsRequest.newBuilder()
            .setAlbumId(albumId)
            .setLimit(limit)
            .setCursorFileId(cursorFileId)
        if (cursorAddedAtMillis != null) req.setCursorAddedAt(cursorAddedAtMillis.toTimestamp())
        when (kindFilter) {
            CloudMediaKind.VIDEO -> req.setKindFilter(MediaKind.MEDIA_KIND_VIDEO)
            CloudMediaKind.PHOTO -> req.setKindFilter(MediaKind.MEDIA_KIND_PHOTO)
            else -> Unit
        }
        val resp = grpc.albumStub().listAlbumItems(req.build())
        return AlbumItemsPage(
            items = resp.itemsList.map { MediaAsset.from(it.file) },
            nextCursorAddedAtMillis = if (resp.hasNextCursorAddedAt()) resp.nextCursorAddedAt.toEpochMillis() else null,
            nextCursorFileId = resp.nextCursorFileId,
        )
    }

    suspend fun createAlbum(name: String, description: String = ""): AlbumCard {
        val resp = grpc.albumStub().createAlbum(
            CreateAlbumRequest.newBuilder().setName(name).setDescription(description).build()
        )
        return AlbumCard.from(resp)
    }

    suspend fun updateAlbum(
        albumId: String,
        name: String? = null,
        description: String? = null,
        coverFileId: String? = null,
    ): AlbumCard {
        val req = UpdateAlbumRequest.newBuilder().setAlbumId(albumId)
        if (name != null) req.setName(name)
        if (description != null) req.setDescription(description)
        if (coverFileId != null) req.setCoverFileId(coverFileId)
        return AlbumCard.from(grpc.albumStub().updateAlbum(req.build()))
    }

    suspend fun deleteAlbum(albumId: String) {
        grpc.albumStub().deleteAlbum(DeleteAlbumRequest.newBuilder().setAlbumId(albumId).build())
    }

    suspend fun addItems(albumId: String, fileIds: List<String>) {
        grpc.albumStub().addItemsToAlbum(
            AddItemsToAlbumRequest.newBuilder().setAlbumId(albumId).addAllFileIds(fileIds).build()
        )
    }

    suspend fun removeItems(albumId: String, fileIds: List<String>) {
        grpc.albumStub().removeItemsFromAlbum(
            RemoveItemsFromAlbumRequest.newBuilder().setAlbumId(albumId).addAllFileIds(fileIds).build()
        )
    }
}

File: Mac/BarkCloudKit/Sources/BarkCloudKit/Data/Cloud/AlbumRepository.swift
import Foundation
import GRPCCore
import SwiftProtobuf

/// Доступ к сервису альбомов (`AlbumApi`). Доменные ошибки пробрасываются как
/// `RPCError`; UI маппит их через `domainErrorMessage(_:)`.
public final class AlbumRepository: Sendable {
    private let grpc: GrpcManager

    public init(grpc: GrpcManager) { self.grpc = grpc }

    public func listAlbums(
        limit: Int32 = 50,
        cursorUpdatedAt: Date? = nil,
        cursorAlbumID: String = ""
    ) async throws -> AlbumPage {
        let stub = try await grpc.albumStub()
        var req = Barkcloud_Files_ListAlbumsRequest()
        req.limit = limit
        if let cursorUpdatedAt {
            req.cursorUpdatedAt = Google_Protobuf_Timestamp(date: cursorUpdatedAt)
        }
        req.cursorAlbumID = cursorAlbumID
        let resp = try await stub.listAlbums(req)
        return AlbumPage(
            albums: resp.albums.map(AlbumCard.init),
            nextCursorUpdatedAt: resp.hasNextCursorUpdatedAt ? resp.nextCursorUpdatedAt.date : nil,
            nextCursorAlbumID: resp.nextCursorAlbumID
        )
    }

    /// Элементы альбома (с опциональным фильтром по типу).
    public func listItems(
        albumID: String,
        kindFilter: CloudMediaKind? = nil,
        limit: Int32 = 50,
        cursorAddedAt: Date? = nil,
        cursorFileID: String = ""
    ) async throws -> (items: [MediaAsset], nextAddedAt: Date?, nextFileID: String) {
        let stub = try await grpc.albumStub()
        var req = Barkcloud_Files_ListAlbumItemsRequest()
        req.albumID = albumID
        req.limit = limit
        if let cursorAddedAt {
            req.cursorAddedAt = Google_Protobuf_Timestamp(date: cursorAddedAt)
        }
        req.cursorFileID = cursorFileID
        if kindFilter == .video {
            req.kindFilter = .video
        } else if kindFilter == .photo {
            req.kindFilter = .photo
        }
        let resp = try await stub.listAlbumItems(req)
        return (
            resp.items.map { MediaAsset($0.file) },
            resp.hasNextCursorAddedAt ? resp.nextCursorAddedAt.date : nil,
            resp.nextCursorFileID
        )
    }

    @discardableResult
    public func createAlbum(name: String, description: String = "") async throws -> AlbumCard {
        let stub = try await grpc.albumStub()
        var req = Barkcloud_Files_CreateAlbumRequest()
        req.name = name
        req.description_p = description
        return AlbumCard(try await stub.createAlbum(req))
    }

    @discardableResult
    public func updateAlbum(
        albumID: String,
        name: String? = nil,
        description: String? = nil,
        coverFileID: String? = nil
    ) async throws -> AlbumCard {
        let stub = try await grpc.albumStub()
        var req = Barkcloud_Files_UpdateAlbumRequest()
        req.albumID = albumID
        if let name { req.name = name }
        if let description { req.description_p = description }
        if let coverFileID { req.coverFileID = coverFileID }
        return AlbumCard(try await stub.updateAlbum(req))
    }

    public func deleteAlbum(_ albumID: String) async throws {
        let stub = try await grpc.albumStub()
        var req = Barkcloud_Files_DeleteAlbumRequest()
        req.albumID = albumID
        _ = try await stub.deleteAlbum(req)
    }

    public func addItems(albumID: String, fileIDs: [String]) async throws {
        let stub = try await grpc.albumStub()
        var req = Barkcloud_Files_AddItemsToAlbumRequest()
        req.albumID = albumID
        req.fileIds = fileIDs
        _ = try await stub.addItemsToAlbum(req)
    }

    public func removeItems(albumID: String, fileIDs: [String]) async throws {
        let stub = try await grpc.albumStub()
        var req = Barkcloud_Files_RemoveItemsFromAlbumRequest()
        req.albumID = albumID
        req.fileIds = fileIDs
        _ = try await stub.removeItemsFromAlbum(req)
    }
}
